### PR TITLE
Add support for Endpoint's preferred IPv6 address

### PR DESCRIPTION
**Pull request by aboch**
Signed-off-by: Alessandro Boch <aboch@docker.com>

### CODE REVIEW

File: endpoint.go [hunk 1]
```diff
@@ -61,6 +61,7 @@ type endpoint struct {
 	generic       map[string]interface{}
 	joinLeaveDone chan struct{}
 	prefAddress   net.IP
+	prefAddressV6 net.IP
```
Comment (mavenugo) [inline endpoint.go]: I think we should treat it as a multiple ip-address requirement that @jc-m raised.
That will cover both ipv4 and ipv6 cases. WDYT ?

Comment (aboch) [inline endpoint.go]: I thought of that, but I felt it is better to have two separate fields for v4 and v6 otherwise we have to go through the list each time to find out whether an ipv6 was specified.

Also I feel it is better to keep the primary ip separated from the list of secondary: Actually in the code @jc-m shared with us, he is also keeping the secondary IPs in a separate list.
I agree with that, also because we may not want to enforce the <ip belongs to the subnet> for secondary ip and also discard the possible <already allocated error> to cover the use case where the secondary IP is a VIP shared by more than one container.

In summary, the allocation of the secondary IPs would likely take a different path than the allocation of the primary, therefore it is not ideal to keep all of them in the same list, even more to only have one list for v4 and v6.

Comment (mavenugo) [inline endpoint.go]: @mrjana WDYT ?

Comment (mrjana) [inline endpoint.go]: +1 on keeping them separate because these are primary IP addresses of the container and primary IP addresses have a special meaning in that they have to be unique and so libnetwork needs to handle them that way and make appropriate IPAM calls to satisfy that requirement. In the future if we try to provide multiple IP addresses then we need to think about their characteristics and how libnetwork should handle them. But that can be kept separate from this discussion and PR.

Comment (mavenugo) [inline endpoint.go]: okay. thanks.

File: endpoint.go [hunk 2]
```diff
@@ -804,9 +810,27 @@ func (ep *endpoint) assignAddressVersion(ipVer int, ipam ipamapi.Ipam) error {
 		return nil
 	}
 
+	// Validate preferred address
+	if prefAdd != nil {
+		for _, d := range ipInfo {
```
Comment (mrjana) [inline endpoint.go]: Why do we need this extra iteration for validation? Why can't we do this validation in the iterator below which anyways had to iterate over `ipInfo`? Reduces some extra boiler plate code

Comment (aboch) [inline endpoint.go]: At first I was not able to combine both things in the following loop in a clean way.
Let me give it one more try.

**Comment by aboch** [comment]
Thanks @mrjana for the comment. Took care of it.
PTAL.

**Comment by mrjana** [comment]
LGTM

**Comment by mavenugo** [comment]
LGTM
